gkctl/main.c

```diff
@@ -0,0 +1,252 @@
+load_file_to_buffer(const char *file_name, char *buffer, int n)
+{
+	int ret = 0;
+	long fsize;
+	FILE *fp;
+
+	memset(buffer, 0, n);
+
+	if ((fp = fopen(file_name, "r")) == NULL) {
+		fprintf(stderr, "Error: Failed to open file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	ret = fseek(fp, 0, SEEK_END);
+	if (ret < 0) {
+		fprintf(stderr, "Error: Failed to seek to the end of file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto close_file;
+	}
+
+	fsize = ftell(fp);
+	if (fsize < 0) {
+		fprintf(stderr, "Error: Failed to obtain the current value of the file position indicator for file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto close_file;
+	} else if (fsize > n) {
+		fprintf(stderr, "Error: file %s is too long (length: %ld), so cannot be fitted into the buffer with size %d\n",
+			file_name, fsize, n);
+		ret = -1;
+		goto close_file;
+	}
+
+	ret = fseek(fp, 0, SEEK_SET);
+	if (ret < 0) {
+		fprintf(stderr, "Error: Failed to seek to the start of file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto close_file;
+	}
+
+	/*
+	 * Return value equals the number of bytes transferred
+	 * only when size (i.e., second parameter) is 1.
+	 */
+	ret = fread(buffer, 1, fsize, fp);
+	if (ret != fsize) {
+		fprintf(stderr, "Error: Failed to read the whole file %s\n",
+			file_name);
+		if (ferror(fp))
+			fprintf(stderr, "Error: %s\n", strerror(errno));
+		ret = -1;
+	}
+
+close_file:
+	fclose(fp);
+out:
+	return ret;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	char send_buff[MSG_MAX_LEN + 2];
+	char recv_buff[MSG_MAX_LEN + 2];
+	size_t len = 0;
+	size_t total_file_len = 0;
+	int send_size = 0;
+	int recv_size = 0;
+	int tot_send = 0;
+	int tot_recv = 0;
+	struct sockaddr_un serv_addr; 
+
+    	struct args args = {
+		/* Defaults. */
+		.server_path = "/tmp/dyn_cfg.socket",
+	};
+
+	/* Read parameters. */
+	argp_parse(&argp, argc, argv, 0, NULL, &args);
+
+	if (sizeof(serv_addr.sun_path) <= strlen(args.server_path)) {
+		fprintf(stderr, "Error: passing a too long server path (i.e., > %lu) - %s!\n",
+			sizeof(serv_addr.sun_path), args.server_path);
+		ret = -1;
+		goto out;
+	}
+	serv_addr.sun_family = AF_UNIX;
+	strncpy(serv_addr.sun_path, args.server_path,
+		strlen(args.server_path));
+	serv_addr.sun_path[strlen(args.server_path)] = '\0';
+
+	ret = load_file_to_buffer(args.lua_script_path,
+		send_buff + sizeof(uint16_t), MSG_MAX_LEN);
+	if (ret <= 0) {
+		ret = -1;
+		goto out;
+	}
+	total_file_len = ret;
+	*(uint16_t *)send_buff = htons(ret);
+
+	if((sock_fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
+		perror("Error : Could not create socket\n");
+		ret = -1;
+		goto out;
+	}
+
+	if(connect(sock_fd, (struct sockaddr *)&serv_addr,
+			sizeof(serv_addr)) < 0) {
+		perror("Error : Connect failed\n");
+		ret = -1;
+		goto close_sock;
+	}
+
+	len = total_file_len + sizeof(uint16_t);
+	while ((send_size = send(sock_fd, send_buff + tot_send,
```

Comment:
`send()` may return an error; handle it properly. Encapsulate this block of code into a function: `write_all()`.